### DISCUSSION TITLE
Deconstruct with exact type match should be preferred in case in ambiguity

### DISCUSSION THREAD

**Discussion by Happypig375** [dotnet/csharplang#3828] [Language Ideas]
**Version Used**: latest master (687393280df246e67b5878c5377b1a6b062fa3b1)

**Steps to Reproduce**: https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4DsJjlADhAxnAAgAEBGAOgCUBXFGASwzIGEB7MLOgG0QGVEA3OvigEA3sQBMBAJJQAogA94CNJ2ko6MMQQC+ugLAAoI0QDMBBHFwsEAEwIANABR1aBBQEptOk+cvW7AgBNJwd3L3EzYgAWAgARKxYUWAQqXBgnFiotVy0wLwBeAD4CMAIC9zIFAG4DY0MiKQAtbWAWFk4CACEnIIIAT0KS/oI6ERCSD1qfQyA==

```cs
namespace System.Runtime.CompilerServices { class IsExternalInit { } }

public record X(int x) { }
public record Y(X x) { public void Deconstruct(out int m) => m = x.x; }

class Z { bool B(Y y) => y is Y(1); }
```

**Expected Behavior**: The overload with `int` is chosen since `1` is an `int` which is an exact type match.

**Actual Behavior**: `error CS0121: The call is ambiguous between the following methods or properties: 'Y.Deconstruct(out X)' and 'Y.Deconstruct(out int)'`

**Comment by gafter** [comment]
Since this is requesting a language change, transferring to `csharplang`.

**Comment by HaloFour** [comment]
https://github.com/dotnet/csharplang/issues/1998#issuecomment-438434651
https://github.com/dotnet/csharplang/issues/1998#issuecomment-438472660

**Comment by Happypig375** [comment]
@HaloFour Well there's no way to disable the generated Deconstruct with record types without losing the convenient shorthand for properties and constructors.

**Comment by HaloFour** [comment]
@Happypig375 

Symmetry between construction/deconstruction is one of the purposes of positional records.  You could use a nominal record.  That doesn't have either a constructor or deconstructor, but you can use initializers and bring your own Deconstruct method.
